monai/apps/pathology/transforms/post/array.py

```diff
@@ -68,56 +69,70 @@ def __init__(
     def __call__(self, prob_map: NdarrayOrTensor, hover_map: NdarrayOrTensor) -> NdarrayOrTensor:  # type: ignore
         """
         Args:
-            prob_map: the probability map output of the NP branch, shape must be [1, H, W, [D]].
+            prob_map: the probability map output of the NP branch, shape must be [1, H, W].
             hover_map: the horizontal and vertical distances of nuclear pixels to their centres
-                of mass output from the HV branch, shape must be [2, H, W, [D]].
+                of mass output from the HV branch, shape must be [2, H, W].
 
         Returns:
-            instance labelled segmentation map with shape [1, H, W, [D]].
+            instance labelled segmentation map with shape [1, H, W].
+
+        Raises:
+            ValueError: when the `prob_map` dimension is not [1, H, W].
+            ValueError: when the `hover_map` dimension is not [2, H, W].
+
         """
-        pred = convert_to_tensor(prob_map, track_meta=get_track_meta())
+        prob_map = convert_to_tensor(prob_map, track_meta=get_track_meta())
         hover_map = convert_to_tensor(hover_map, track_meta=get_track_meta())
 
+        if prob_map.shape[0] != 1:
+            raise ValueError("Only supports single channel probability map!")
+
+        if hover_map.shape[0] != 2:
+            raise ValueError("Hover map should with shape (2, H, W)!")
+
+        if len(prob_map.shape) > 3 or len(hover_map.shape) > 3:  # only for 2D
+            raise ValueError("Only supports probability map with shape (1, H, W)!")
+
+        pred = convert_to_numpy(prob_map)
+        hover_map = convert_to_numpy(hover_map)
+
         # processing
         pred = self.pred_discreter(pred)
+        pred = label(pred)[0]
         pred = self.remove_small_objects(pred)
+        pred[pred > 0] = 1
 
-        hover_h = hover_map[0:1, ...]
-        hover_v = hover_map[1:2, ...]
-
-        hover_h = (hover_h - torch.min(hover_h)) / (torch.max(hover_h) - torch.min(hover_h))  # type: ignore
-        hover_v = (hover_v - torch.min(hover_v)) / (torch.max(hover_v) - torch.min(hover_v))  # type: ignore
+        hover_h = hover_map[0, ...]
+        hover_v = hover_map[1, ...]
 
-        sobelh = self.sobel_gradient(hover_h)[0, ...]
-        sobelv = self.sobel_gradient(hover_v)[1, ...]
-        sobelh = 1 - (sobelh - torch.min(sobelh)) / (torch.max(sobelh) - torch.min(sobelh))  # type: ignore
-        sobelv = 1 - (sobelv - torch.min(sobelv)) / (torch.max(sobelv) - torch.min(sobelv))  # type: ignore
+        hover_h = (hover_h - np.min(hover_h)) / (np.max(hover_h) - np.min(hover_h))
+        hover_v = (hover_v - np.min(hover_v)) / (np.max(hover_v) - np.min(hover_v))
+        sobelh = cv2.Sobel(hover_h, cv2.CV_64F, 1, 0, ksize=self.kernel_size)
```

Comment:
What's the reason for using cv2 here? I would have thought that we'd prefer to use the Monai implementation - or the skimage one.

Reply:
Hi @JHancox, when I test `CalcualteInstanceSegmentationMap` with the data you shared with me, I found some [different effects](https://github.com/Project-MONAI/MONAI/pull/5189#issuecomment-1256918303) between cv2 and `SobelGradients`. The difference between cv2 and `SobelGradients` also causes `CalcualteInstanceSegmentationMap` can't get correct instance segmentation results. The results I posted in [this PR](https://github.com/Project-MONAI/MONAI/pull/5189#issuecomment-1257390456) Thanks!

Reply:
Oh, I see. That's odd. Good find! I'll check whether I see similar differences.